#  AI-Powered Retail Intelligence & Demand Forecasting Platform

#  Module 2: Data Cleaning

##  Objective

The objective of this module is to identify and correct data quality issues before performing analysis and machine learning.

Data cleaning improves the reliability, consistency, and accuracy of the dataset.

A clean dataset leads to better visualizations, better insights, and more accurate predictive models.

---

## What is Data Cleaning?

Data Cleaning is the process of identifying and correcting inaccurate, inconsistent, duplicate, incomplete, or improperly formatted data.

Real-world datasets are rarely perfect.

Before analysis or machine learning, a Data Scientist ensures that the dataset is clean and reliable.

Common data cleaning tasks include:

- Handling missing values
- Removing duplicate records
- Correcting data types
- Detecting outliers
- Standardizing inconsistent values
- Validating data integrity

A well-cleaned dataset improves the quality of business insights and predictive models.

---

### Step 1 : Import Libraries

In [17]:
import pandas as pd
import numpy as np

### Step 2 : Load Dataset

In [18]:
df = pd.read_csv("../data/raw/Global_Superstore_2016.csv", encoding="latin1")

### Step 3: Creating backup copy of the dataset

#### Why do we create a copy?

Before modifying a dataset, it is considered a good practice to create a copy.

This preserves the original dataset and allows us to perform cleaning operations safely.

If any mistake occurs during preprocessing, the original data remains unchanged.

Professional Data Scientists rarely modify the original dataset directly.

In [19]:
df_clean = df.copy()

## Data Cleaning Checklist

The following checks will be performed:

- Check dataset dimensions
- Check duplicate records
- Check missing values
- Check incorrect data types
- Convert date columns
- Check inconsistent values
- Detect outliers
- Save cleaned dataset

## Step 1: Check Dataset Shape

The shape of the dataset is checked to determine the total number of rows and columns before performing any cleaning operations.
Code

In [4]:
df.shape

(51290, 24)

## Step 2: Check Missing Values

Missing values reduce data quality and can negatively affect analysis and machine learning models. Therefore, they are identified before cleaning.

In [5]:
df.isnull().sum()

Row ID                0
Order ID              0
Order Date            0
Ship Date             0
Ship Mode             0
Customer ID           0
Customer Name         0
Segment               0
Postal Code       41296
City                  0
State                 0
Country               0
Region                0
Market                0
Product ID            0
Category              0
Sub-Category          0
Product Name          0
Sales                 0
Quantity              0
Discount              0
Profit                0
Shipping Cost         0
Order Priority        0
dtype: int64

### Handle Missing Values

Missing values reduce data quality and may affect analysis and machine learning models.

The dataset was inspected for missing values using the `isnull().sum()` function.

The analysis revealed that only the **Postal Code** column contains missing values. Since more than 80% of its values are missing and the column is not essential for business analysis or demand forecasting, it is removed from the dataset.

In [6]:
df.drop(columns=["Postal Code"], inplace=True)

In [7]:
df.isnull().sum()

Row ID            0
Order ID          0
Order Date        0
Ship Date         0
Ship Mode         0
Customer ID       0
Customer Name     0
Segment           0
City              0
State             0
Country           0
Region            0
Market            0
Product ID        0
Category          0
Sub-Category      0
Product Name      0
Sales             0
Quantity          0
Discount          0
Profit            0
Shipping Cost     0
Order Priority    0
dtype: int64

##### Observation

- Only the Postal Code column contained missing values.
- More than 80% of the values were missing.
- Since the column is not required for sales analysis or machine learning, it was removed.
- The dataset is now free from missing values.

### Step 3: Check Duplicate Records

Duplicate records can lead to inaccurate analysis and biased machine learning models. Therefore, duplicate entries are checked before proceeding further.

In [8]:
df.duplicated().sum()

np.int64(0)

#### Observation

- No duplicate records were found in the dataset.
- This indicates that each transaction is unique.
- Therefore, no duplicate rows were removed.

### Step 4: Check Data Types

Every column in a dataset has a specific data type, such as integer, float, object (text), or datetime.

Checking data types is important because incorrect data types can lead to errors during analysis and machine learning.

For example, date columns stored as text (object) must be converted to datetime format before performing time-based analysis.

In [9]:
df.dtypes

Row ID              int64
Order ID              str
Order Date            str
Ship Date             str
Ship Mode             str
Customer ID           str
Customer Name         str
Segment               str
City                  str
State                 str
Country               str
Region                str
Market                str
Product ID            str
Category              str
Sub-Category          str
Product Name          str
Sales             float64
Quantity            int64
Discount          float64
Profit            float64
Shipping Cost     float64
Order Priority        str
dtype: object

### Step 4: Convert Date Columns

The **Order Date** and **Ship Date** columns are currently stored as text (string/object).

Date columns should be converted into the **datetime** data type because datetime provides several useful functionalities, such as:

- Extracting year, month, day, and quarter
- Performing monthly and yearly sales analysis
- Creating time series visualizations
- Building demand forecasting models

Keeping date columns as text limits analytical capabilities and may lead to errors during time-based analysis.

In [20]:
df["Order Date"] = pd.to_datetime(df_clean["Order Date"])
df["Ship Date"] = pd.to_datetime(df_clean["Ship Date"])

In [21]:
df.dtypes


Row ID                     int64
Order ID                     str
Order Date        datetime64[us]
Ship Date         datetime64[us]
Ship Mode                    str
Customer ID                  str
Customer Name                str
Segment                      str
Postal Code              float64
City                         str
State                        str
Country                      str
Region                       str
Market                       str
Product ID                   str
Category                     str
Sub-Category                 str
Product Name                 str
Sales                    float64
Quantity                   int64
Discount                 float64
Profit                   float64
Shipping Cost            float64
Order Priority               str
dtype: object

#### Observation

The **Order Date** and **Ship Date** columns were successfully converted from string/object to datetime format.

##### Action Taken

The `pd.to_datetime()` function was used to convert both date columns.

##### Business Impact

The dataset is now ready for time-based analysis, including monthly sales trends, seasonal analysis, and demand forecasting.

### Step 5: Validate Numerical Columns

Before proceeding with analysis, numerical columns are inspected to identify any invalid or unrealistic values.

This helps ensure the dataset is logically consistent and suitable for business analysis.

In [12]:
df.describe()

,Row ID,Order Date,Ship Date,Sales,Quantity,Discount,Profit,Shipping Cost
count,51290.00000,51290,51290,51290.000000,51290.000000,51290.000000,51290.000000,51290.000000
mean,25645.50000,2014-05-11 21:26:49.155780,2014-05-15 20:42:42.745174,246.490581,3.476545,0.142908,28.610982,26.478567
min,1.00000,2012-01-01 00:00:00,2012-01-03 00:00:00,0.444000,1.000000,0.000000,-6599.978000,1.002000
25%,12823.25000,2013-06-19 00:00:00,2013-06-23 00:00:00,30.758625,2.000000,0.000000,0.000000,2.610000
50%,25645.50000,2014-07-08 00:00:00,2014-07-12 00:00:00,85.053000,3.000000,0.000000,9.240000,7.790000
75%,38467.75000,2015-05-22 00:00:00,2015-05-26 00:00:00,251.053200,5.000000,0.200000,36.810000,24.450000
max,51290.00000,2015-12-31 00:00:00,2016-01-07 00:00:00,22638.480000,14.000000,0.850000,8399.976000,933.570000
std,14806.29199,NaN,NaN,487.565361,2.278766,0.212280,174.340972,57.251373


In [22]:
# 1. Date columns ko Datetime format me convert karna
df['Order Date'] = pd.to_datetime(df['Order Date'])
df['Ship Date'] = pd.to_datetime(df['Ship Date'])

# 2. Postal Code column drop karna (80%+ null values + City/State/Country already context de rahe hain)
df.drop(columns=['Postal Code'], inplace=True)

# 3. Column Names ko standardize karna (lowercase + spaces ko '_' se replace karna)
# Isse aage coding aur indexing bahut easy ho jayegi
df.columns = df.columns.str.lower().str.replace(' ', '_').str.replace('-', '_')

# 4. Preprocessing Sanity Checks
print("--- Data Cleaning Summary ---")
print("New Shape:", df.shape)
print("Remaining Null Values:\n", df.isnull().sum().sum())
print("\nUpdated Data Types:")
print(df.dtypes[['order_date', 'ship_date']])

--- Data Cleaning Summary ---
New Shape: (51290, 23)
Remaining Null Values:
 0

Updated Data Types:
order_date    datetime64[us]
ship_date     datetime64[us]
dtype: object


#### Observation

- Sales values range from **0.44** to **22,638.48**, indicating a wide variation in transaction values.
- Quantity values range from **1** to **14**, which appears reasonable.
- Discounts range from **0%** to **85%**, suggesting that some products receive very high discounts.
- Profit contains both positive and negative values, indicating that some transactions resulted in losses.
- Shipping Cost values appear valid and contain no obvious inconsistencies.

#### Action Taken

The numerical columns were inspected for logical consistency.

No invalid numerical values were found at this stage.

#### Business Impact

The dataset contains realistic numerical values and is suitable for further exploratory analysis.

The presence of negative profits and high discounts provides opportunities for deeper business insights during the EDA phase.

In [24]:
import pandas as pd
import numpy as np

# 1. Load Data
df = pd.read_csv("../data/raw/Global_Superstore_2016.csv", encoding="latin1")

# 2. Drop high null column
if "Postal Code" in df.columns:
    df.drop(columns=["Postal Code"], inplace=True)

# 3. Standardize Column Names (snake_case)
df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_').str.replace('-', '_')

# 4. Fix Datatypes
df['order_date'] = pd.to_datetime(df['order_date'])
df['ship_date'] = pd.to_datetime(df['ship_date'])

# 5. Sanity Check 1: Strip extra spaces from string columns
str_cols = df.select_dtypes(include=['object', 'string']).columns
for col in str_cols:
    df[col] = df[col].astype(str).str.strip()

# 6. Sanity Check 2: Shipping Days Feature (Model ke liye helpful hoga)
df['shipping_days'] = (df['ship_date'] - df['order_date']).dt.days

# Filter invalid transactions (agar shipping days negative hain)
df = df[df['shipping_days'] >= 0]

# 7. Sanity Check 3: Quantity aur Sales positive honi chahiye
df = df[(df['sales'] > 0) & (df['quantity'] > 0)]

# Summary Verification
print("--- CLEAN DATA SUMMARY ---")
print("Shape:", df.shape)
print("Null Values Remaining:", df.isnull().sum().sum())
print("Date Range:", df['order_date'].min(), "to", df['order_date'].max())
print("\nSample Clean Data:")
print(df.head(2))

# 8. Save CSV Correctly (without 'df =')
df.to_csv("../data/processed/cleaned_superstore.csv", index=False)
print("\nSuccessfully saved to ../data/processed/cleaned_superstore.csv")

--- CLEAN DATA SUMMARY ---
Shape: (51290, 24)
Null Values Remaining: 0
Date Range: 2012-01-01 00:00:00 to 2015-12-31 00:00:00

Sample Clean Data:
   row_id                  order_id order_date  ship_date     ship_mode  \
0   40098  CA-2014-AB10015140-41954 2014-11-11 2014-11-13   First Class   
1   26341    IN-2014-JR162107-41675 2014-02-05 2014-02-07  Second Class   

    customer_id  customer_name    segment           city            state  \
0  AB-100151402  Aaron Bergman   Consumer  Oklahoma City         Oklahoma   
1     JR-162107  Justin Ritter  Corporate     Wollongong  New South Wales   

   ...    category sub_category                               product_name  \
0  ...  Technology       Phones                           Samsung Convoy 3   
1  ...   Furniture       Chairs  Novimex Executive Leather Armchair, Black   

      sales quantity discount    profit  shipping_cost  order_priority  \
0   221.980        2      0.0   62.1544          40.77            High   
1  3709.395  

## The cleaned dataset is saved in processed folder of data 
